Fix stack trace output for TestNG failures

Kropie opened the pull request:
This will fixes #1260.

jdneo commented:
Thank you for your contribution, I'll take a look this week

jdneo commented:
I tested the PR locally, but seems it's still not showing the errors according to the steps in #1260? Did I miss anything?

BTW, the refactoring looks good.

jdneo commented on src/runners/testngRunner/TestNGRunnerResultAnalyzer.ts:
```diff
@@ -40,7 +41,7 @@ export class TestNGRunnerResultAnalyzer implements IRunnerResultAnalyzer {
 
     public analyzeData(data: string): void {
-        const lines: string[] = data.split(/\r?\n/);
+        const lines: string[] = this.unescape(data).split(/\r?\n/);
```
I found the code in `analyzeData()` has some bugs, which I changed it to:

```js
public analyzeData(data: string): void {
    let match: RegExpExecArray | null;
    // tslint:disable-next-line: no-conditional-assignment
    while ((match = this.regex.exec(data)) !== null) {
        try {
           this.processData(match[1]);
        } catch (error) {
            this.testContext.testRun.appendOutput(`[ERROR] Failed to parse output data: ${match[1]}\n`);
        }
    }
}
```

Meanwhile, add the `g` flag to the regexp @ line 16:

```js
private readonly regex: RegExp = /@@<TestRunner-({[\s\S]*?})-TestRunner>/g;
```

jdneo replied:
Ah, I see why it's not working. Because here `.unescape(data)` is called before splitting to lines. Is there any reason to do this?

jdneo commented:
@Kropie Would you like to update your PR?

Kropie commented:
@jdneo I will update my pull request tomorrow morning with the recommendations that you provided above. Thanks for your help on this!

Kropie commented:
@jdneo  I've updated this PR to address the comments that you have above. Additionally I cleaned up the failure processing logic to clean up the "peek" view for test failures.

Kropie commented:
This is the result of the latest commit when running a failing test.
![image](https://user-images.githubusercontent.com/11445940/130242485-c5d255ef-2c33-4a38-b336-b353d81c7d5b.png)

jdneo commented:
@Kropie The PR looks good now. the common logic is shared with this refactoring. But seems that it does not fix #1260, did I miss anything?

#1260 is a bug that when the suite is failed to run, nothing shows in UI except for the output channel. You can try download the project attached in #1260, and click run all.

Kropie commented:
@jdneo I may have missed a portion of the problem statement for #1260, but this does solve the formatting issues that were previously present when you would print out the results from the "peek error" screen. Additionally this resolves the stack trace issues in the preview.

Should the stack trace also be printing out on the debug console?

Kropie commented:
After looking running with the code linked in the pull request, this PR will resolve the formatting issues. That being said, if an exception is thrown while the tests are running all subsequent tests will not be able to run.

Kropie commented:
This is what I'm seeing when I run all the tests in that sample project
![image](https://user-images.githubusercontent.com/11445940/130309344-22795a76-5b3f-42d3-a378-c490668aba52.png)

Kropie commented:
Finally figured out what I believe is the root of the problem. The tests in the example project use groups, but the test runner currently does not support selenium groups. I just pushed another commit that logs the error clearly to the user.

I think with the additional logging we have addressed the original issue in #1260 and an additional ticket should be written for adding TestNG group support to the java test runner extension.

Now when you run the tests you will see the following error message in the debug console.

Kropie commented:
This is an example of an error that is logged to the console to make things clearer to the user
![image](https://user-images.githubusercontent.com/11445940/130312709-3889eb3f-58e9-4a45-9a7b-671a2b928285.png)

jdneo commented:
The screenshot looks good. I'll take a deep look tomorrow.

> I think with the additional logging we have addressed the original issue in #1260 and an additional ticket should be written for adding TestNG group support to the java test runner extension.

Yes as you said, the current implementation of the TestNG runner has a lot of limitations. I'm hoping to migrate to the TestNG's official runner: https://github.com/testng-team/testng-remote, but so far does not have time to do this. 😥

jdneo commented on src/runners/testngRunner/TestNGRunnerResultAnalyzer.ts:
```diff
@@ -83,35 +79,15 @@ export class TestNGRunnerResultAnalyzer implements IRunnerResultAnalyzer {
-                    testMessage.location = new Location(item.uri, item.range);
+
+                for (const line of outputData.attributes.trace.split(/\r?\n/)) {
+                    this.processStackTrace(line, markdownTrace, testMessage, this.currentItem, this.projectName);
```
Here `testMessage` is passed as `assertionFailure` but actually it's not.

`assertionFailure` is the test message looks like this:
![image](https://user-images.githubusercontent.com/6193897/130346452-40616b6b-9850-4aef-b0c4-b7090e690c22.png)

you can see it's not a stack trace actually.

I guess that's why the stack trace now only have just few lines.

Kropie replied:
Okay, @jdneo I believe I've resolved the issue with the stack trace that is being presented in the error message. I'm not positive what exactly it is supposed to look like but I've come up with what I believe is the right information.

Kropie commented:
@jdneo Other than merging in the latest changes in from master is there anything else that needs to be done to merge and close this PR?

jdneo commented:
As I commented above, the refactor introduced a regression for the stacktrace of the TestNG cases. See: https://github.com/microsoft/vscode-java-test/pull/1286#discussion_r693457884

That needs to be resolved before merging.

Kropie commented:
I’ll resolve that tomorrow, thanks!

Kropie commented:
> As I commented above, the refactor introduced a regression for the stacktrace of the TestNG cases. See: [#1286 (comment)](https://github.com/microsoft/vscode-java-test/pull/1286#discussion_r693457884)
> 
> That needs to be resolved before merging.

Okay, @jdneo I believe I've resolved the issue with the stack trace that is being presented in the error message. I'm not positive what exactly it is supposed to look like but I've come up with what I believe is the right information.
### **JUnit**
![image](https://user-images.githubusercontent.com/11445940/130709329-d276a862-6403-4ebc-8492-8be93cbe6eb3.png)
### **TestNG**
![image](https://user-images.githubusercontent.com/11445940/130709511-e49a88d7-a710-4a2c-9470-ac94b2514738.png)

jdneo commented:
I just tried with the new commit but I'm afraid it introduced new regression.

Previously, For JUnit, we can show the assertion failure via the diff view:
![WeChat Screenshot_20210825094934](https://user-images.githubusercontent.com/6193897/130714237-6f83f496-86c0-4928-a5dd-8f61bb256402.png)

But now, it becomes to a plain text with a single line stack trace:
![WeChat Screenshot_20210825094856](https://user-images.githubusercontent.com/6193897/130714286-fa7d01b2-20ef-4c6b-a02b-2ac89e733c4b.png)

I guess you thought it a little bit complicated, base on this [commit](https://github.com/microsoft/vscode-java-test/pull/1286/commits/30565fe486b83b9f3a17f0b7221a1ed80353c79c), you can simply make the following change to make it work:

In TestNGRunnerResultAnalyzer, in the `else if (outputData.name === TEST_FAIL)` block :

```ts
...
if (outputData.attributes.trace) {
    ...
    for (const line of outputData.attributes.trace.split(/\r?\n/)) {
        this.processStackTrace(line, markdownTrace, undefined, this.currentItem, this.projectName);
    }

    const testMessage: TestMessage = new TestMessage(markdownTrace);
    if (item.uri && item.range) {
        testMessage.location = new Location(item.uri, item.range);
    }
    testMessages.push(testMessage);
}
const duration: number = Number.parseInt(outputData.attributes.duration, 10);
setTestState(this.testContext.testRun, item, this.currentTestState, testMessages, duration);
...
```

Just to provide more background information. We have two different test message:
1. The stack trace, which is the one you are dealing with
2. The assertion failure(expected ... but actual ...) showing with a diff view. This is the third argument of your extracted function `processStackTrace`. This is optional, in TestNG, we didn't support it, so you can just pass `undefined` to it.

> meanwhile, it's great that you also extracted the code to set the location of the test messages, but that can be addressed in another PR (to avoid this one become bigger and bigger endlessly).

Kropie commented:
> I just tried with the new commit but I'm afraid it introduced new regression.
> 
> Previously, For JUnit, we can show the assertion failure via the diff view:
> ![WeChat Screenshot_20210825094934](https://user-images.githubusercontent.com/6193897/130714237-6f83f496-86c0-4928-a5dd-8f61bb256402.png)
> 
> But now, it becomes to a plain text with a single line stack trace:
> ![WeChat Screenshot_20210825094856](https://user-images.githubusercontent.com/6193897/130714286-fa7d01b2-20ef-4c6b-a02b-2ac89e733c4b.png)
> 
> I guess you thought it a little bit complicated, base on this [commit](https://github.com/microsoft/vscode-java-test/pull/1286/commits/30565fe486b83b9f3a17f0b7221a1ed80353c79c), you can simply make the following change to make it work:
> 
> In TestNGRunnerResultAnalyzer, in the `else if (outputData.name === TEST_FAIL)` block :
> 
> ```ts
> ...
> if (outputData.attributes.trace) {
>     ...
>     for (const line of outputData.attributes.trace.split(/\r?\n/)) {
>         this.processStackTrace(line, markdownTrace, undefined, this.currentItem, this.projectName);
>     }
> 
>     const testMessage: TestMessage = new TestMessage(markdownTrace);
>     if (item.uri && item.range) {
>         testMessage.location = new Location(item.uri, item.range);
>     }
>     testMessages.push(testMessage);
> }
> const duration: number = Number.parseInt(outputData.attributes.duration, 10);
> setTestState(this.testContext.testRun, item, this.currentTestState, testMessages, duration);
> ...
> ```
> 
> Just to provide more background information. We have two different test message:
> 
> 1. The stack trace, which is the one you are dealing with
> 2. The assertion failure(expected ... but actual ...) showing with a diff view. This is the third argument of your extracted function `processStackTrace`. This is optional, in TestNG, we didn't support it, so you can just pass `undefined` to it.
> 
> > meanwhile, it's great that you also extracted the code to set the location of the test messages, but that can be addressed in another PR (to avoid this one become bigger and bigger endlessly).

First, thanks for your patience!
Second, I was able to resolve the regression with the "Expected ... but actual ... logic" that you referenced above.

I pushed the update out based on my last commit because the delta was rather small. Let me know if you would still rather have me do the code update that you referenced above on the previous commit.

jdneo commented:
Thanks for the quite response! Personally, I would prefer to reset to the commit I mentioned and append commit based on that. (To make the total change as simple as possible). We can definitely address other issues by several PRs separately.

Kropie commented:
> Thanks for the quite response! Personally, I would prefer to reset to the commit I mentioned and append commit based on that. (To make the total change as simple as possible). We can definitely address other issues by several PRs separately.

Just tried to make the modification to the [commit](https://github.com/microsoft/vscode-java-test/pull/1286/commits/30565fe486b83b9f3a17f0b7221a1ed80353c79c) referenced above, but that has issues that are addressed by my latest commit. Specifically if you have a stack trace with multiple elements in it and you have a failure message instead of a "expected .. but actual ..." message you will only see the first element on the stack at the time of the failure.
<br>
Additionally for TestNG tests you only ever get the first element on the stack at the time of failure.
<br>
Also, for help visualizing the problem:
### Limited Stack Trace (Update of referenced [commit](https://github.com/microsoft/vscode-java-test/pull/1286/commits/30565fe486b83b9f3a17f0b7221a1ed80353c79c))
![image](https://user-images.githubusercontent.com/11445940/130719903-5c1899b4-5263-4434-97e0-f1f724440680.png)

### Full Stack Trace (Current latest commit in this PR)
![image](https://user-images.githubusercontent.com/11445940/130720191-9f03f23d-bf6a-41f1-bfc9-0a7205dff32d.png)

Kropie commented:
That being said I am okay with updating that previous commit and just creating another PR if that is your preference.

Kropie commented:
That being said I am okay with updating that previous commit and just creating another PR if that is your preference.

jdneo commented:
I prefer to resolve them separately in differently. So this PR is aiming to resolve #1283 & #1291.

I believe the problem you mentioned is similar to #1266, and we can for sure resolve that problem in another PR.

Please correct me if I misunderstand

Kropie commented:
That works for me. I'll have the pull request updated tomorrow

jdneo commented:
@Kropie, looks pretty good! Thank you for the nice job!

Kropie commented:
> @Kropie, looks pretty good! Thank you for the nice job!

Thanks for the guidance!
